run-csharp

```diff
@@ -2,10 +2,71 @@
 
 . ./config
 
+CSHARP_TEST_FORMATS_DIR=$FORMATS_COMPILED_DIR/csharp
 CSHARP_SPEC_DIR=spec/csharp
+CSHARP_TEST_OUT_DIR="$TEST_OUT_DIR"/csharp
 
-# Compile everything
-xbuild "$CSHARP_SPEC_DIR/kaitai_struct_csharp_tests.sln"
+rm -rf "$CSHARP_TEST_OUT_DIR"
+mkdir -p "$CSHARP_TEST_OUT_DIR"
+
+# Try and build the project first - if there's no compile errors, we can continue straight away
+echo "run-csharp: building project"
+xbuild "$CSHARP_SPEC_DIR/kaitai_struct_csharp_tests.sln" >> "$CSHARP_TEST_OUT_DIR/_compile.log" || {
+
+    # Some formats failed to build - find the failing files and erase them
+	echo "run-csharp: some formats failed to compile, checking each file individually"
+    
+    # Find the invalid files and wipe them - we do this by scanning the error list from the compiler
+    mcs \
+        -t:library \
+        -fullpaths \
+        -o:$CSHARP_TEST_OUT_DIR/_cs_compile_temp.dll \
+        -r:"$CSHARP_SPEC_DIR/packages/NUnit.3.4.1/lib/net45/nunit.framework.dll" \
+        "$CSHARP_RUNTIME_DIR/*.cs" \
+        "$CSHARP_SPEC_DIR/kaitai_struct_csharp_tests/CommonSpec.cs" \
+        "$CSHARP_TEST_FORMATS_DIR"/*.cs \
+            2>&1 >> "$CSHARP_TEST_OUT_DIR/_compile.log" | \
+            while read -r line ; do
+                # e.g. path/to/file.cs(1): Unexpected symbol
+                # Grab everything before the first '(' character
+                FILENAME=$(echo $line | cut -d '(' -f 1)
+                if [ -s $FILENAME ]
+                then
+                    NAME=$(basename "$FILENAME")
+                    echo "$NAME failed to compile, removing"
+                    > $FILENAME
+                fi
+            done
```

Comment:
Definitely pushing what I know about shell scripting here! Being a Windows user I'm not too familiar with these commands, I'll look those up and see what I can manage.

Will I still need to `while read -r line` after doing that, to erase the files? Or is there some way I can do that without a loop as well?

Like this (just guessing for now):
```
[compile command] | fgrep '(' | cut -d'(' -f1 | while read -r line ; do
    if [ -s "$line" ]
    then
        NAME=$(basename "$line")
        echo "$NAME failed to compile, removing"
        > $line
    fi
done
```